README
secEnvyronment

    Content:
I. Introducing.
II. Tutorial.
III. API reference.
IV. Developer documentation.


    I. Introducing:

This software is used to simplify ClamAV usage and automatize detect verification using Metadefender service.

Scope:
1. Reduce number of false-positive detection;
2. Simplify user interface;
3. Provide less overloaded detection reports.

Usage cases:
1. Manual or scheduled virus-scanning;
2. Simple IP-abuse scanning.

Limitations:
1. Number of Metadefender scans are limited by one's API key. Free API key is limited to 10 scans per day.


    II. Tutorial:

Installation guide is described in INSTALL file.
Application does not require any special workspace preparations.

To get commands help use:
python3 envy_sec.py --help
    or
python3 envy_sec.py -h

To perform file scan:
python3 envy_sec.py -S path/to/file/to/be/scann.ed 
    or
python envy_sec.py --scan-file C:\PathTo\File\ToBeScann.ed

To perform IP scan:
python3 envy_sec.py --scan-ip 8.8.8.8
    or
python3 envy_sec.py -I 8.8.8.0\24

To perform ClamAV signatures update:
python3 envy_sec.py --update
    or
python3 envy_sec.py -U


To send multiple arguments use space:
python3 envy_sec.py -I 8.8.8.8 9.9.9.9


Commands also might be combined:
python3 envy_sec.py --update -I 8.8.8.8 9.9.9.9 -F ./eicar.virus /some/another/file

Command execution priority:
    1. Update;
    2. Scan IP;
    3. Scan file;


    III. API reference:

Source code reference might be found in docs/API reference dir.
(./API-reference/index.html)


    IV. Developer documentation:

This section describes how exactly does secEnvyronment works (or does not). 

    Briefly, how does it work:
User invoke secEnvyronment scan,
secEnvyronment open ClamAV subprocess, 
and put all ClamAV output in queue,
in parallel, thread (work_thread) checks for new lines in queue,
and if thread finds something, it sends it to Metadefender,
then prints out approved or denied result.


                Approve or deny ClamAV detection,
                        output results.
            ________________________________________
           |                                        |
           |                                        |
 (1) secEnvyronment     work_thread ----------> Metadefender
           |                 |        call with
subprocess |                 |      received from
   popen() |                 |        queue args
           |     put in      |      
        ClamAV ----------> queue     


            The main principal work scheme (1 - begin)


    Metadefender Errors:

Table scheme: 
        Category
Response code, Description, HTTP code

1  HTTP code is HTTP status code, used to control HTTP session status 
   (like if server work or not, connection refused and etc.);
2. Response code placed in response JSON, used to control Metadefender session status
   (like data validity, API key limits and etc.);
3. Description is a response code explanation;
   (like why this happened);

        Generic					
400000, Generic error																400
400001, The caching strategy is not recognized										400
400002, The limit strategy is not supported											400
400003, The limit type is not supported												400
400004, The query parameters are not valid											400

        Payload Validation
400020, Header is not valid															400
400021, Body parsing failed															400
400022, Payload validation has failed												400
400023, Headers are not correct														400
400024, Headers are missing															400
400025, Payload is missing or empty													400
400026, Hash in the URL doesn't match the hash value in the body					400
400027, Offset should be a positive integer											400
400028, "Limit should be a positive integer less than 10, 000"						400

        Routing Errors
400040, The requested path is not valid												400
400041, The version is required														400
400042, The version does not exist													400
400043, The requested path does not exist											400
400044, Method does not exist														400
400045, The route was not properly set up											400
400046, The requested route does not exist											400
400047, This route is available only on development environments					400

        Hash Errors	
400060, The `hash` field in the body is required									400
400061, The `hash` field is not an array											400
400062, The `hash` field is empty													400
400063, Exceeded maximum allowed													400
400064, The hash value is not valid													400
400065, The header `include_scan_details` has to be either 0 or 1					400
400066, The header `file_metadata` has to be either 0 or 1							400
400067, Hash update failed															400

        Top Hash Errors
400080, "The amount must be lower than 10, 000 "									400
400081, Type must be one of `clean` / `infected`									400
400082, Period must be one of `day` / `week` / `month`								400
400083, "Threshold must be one of 1, 2, 3, 4, 5, 6 "								400

        appinfo
400100, The fields `os_info.device_identity` are required							400

        Top Detection
400120, The header `x-exclude-empty-file-id` has to be either 0 or 1				400
400121, The header `x-exclude-data` has to be either 0 or 1							400
400122, "The header `x-threshold` must be one of 3, 4, 5, 6 "						400
400123, "Packages should be one of m1, m4, m8, m12, m16, m20, m30 "					400
400124, "Number of hashes must be one of 10, 100, 1, 000, 10, 000 "					400

        Upload Errors
400140, The file upload has failed													400
400141, The header `x-force-scan` has to be either 0 or 1							400
400142, The header `x-sample-sharing` has to be either 0 or 1						400
400143, Private scanning is not enabled for the provided API key					400
400144, Exceeded maximum file size allowed; maximum allowed is 200MB				400
400145, Request body is empty; please send a binary file							400
400146, Provided download URL is not valid or inaccessible							400
400147, Rescan failed. Requested file is missing from our servers.					400
400148, Requested file is a private one and cannot be rescanned						400
400149, Could not update the rescan count											400
400150, The `file_ids` field array in body is required								400
400151, The `file_ids` field is not an array										400
400152, The `file_ids` field is empty												400
400153, Exceeded maximum allowed													400
400154, Exceeded maximum allowed files in archive									400

        API Key Info
400160, The API key you are trying to add already exists							400
400161, The API key could not be removed											400
400162, The API key was not updated													400
400163, The body is invalid															400
400164, No valid operation type														400
400165, No API key specified														400
400166, Please provide a valid email address										400
400167, Please provide a valid body													400
400168, "Please choose another nickname, as this one contains profanities"			400

        IP Scan
400180, Invalid format of input. Provide IPv4 or IPv6.								400
400181, The `ip_addresses` field in body is required								400
400182, The `ip_addresses` field is not an array									400
400183, The `ip_addresses` field is empty											400
400184, Exceeded maximum allowed													400
400185, The address is not a routable IP											400
400186, No response																	400
400187, Invalid response															400

        Stats
400200, The number of days requested must be a positive integer						400
400201, Invalid objectId															400
400202, Invalid date																400
400203, Invalid outbreak report filter												400

        Status
400210, Parameter type must be one of `hashLookup` / `uploadFile` / `ipScan`		400

        Salesforce
400250, Salesforce connectivity error												401
400251, There is no record															401

        Feed
400260, You are allowed to query up to 30 days in the past							400
400261, "Invalid category. Please use: A, D, E, G, M, N, O, P, T, Z "				400

        Authentication
401000, Authentication has failed													401
401001, Authentication strategy is invalid											401
401002, Authentication strategy is not implemented									401
401003, Authorization strategy is not supported for this endpoint					401
401004, Authentication token has expired											401
401005, Authentication token is invalid												401
401006, Invalid API key																401

        Forbidden
403000, Access forbidden															403
403001, Requested resource doesn't match your API key								403
403002, Your IP is blocked because of abuse											403
403003, Insufficient Privileges														403

        Not Found
404000, Endpoint was not found														404
404001, Entity was not found														404
404002, There are no entries found													404
404003, The hash was not found														404
404004, The data_id was not found													404
404005, The hash information was not found											404
404006, There is no data for the selected date										404
404007, Requested file ID does not exist in our records								404
404008, The API key was not found													404

        Payload Acceptance
406000, Content-Type header and payload has to be JSON								406
406001, Payload empty																406

        Request Timeout
408000, Request timeout. It has reached the 60 seconds limit.						408

        Rate Limiting
429000, API key limit exceeded; retry after the limit is reset						429
429001, Your request has been throttled; maximum 10 requests per minute per user	429
429002, Too many connections; try again later										429

        Service Unavailable
503000, External service is not reachable											503
503001, External service is not reachable											503



changelog
0.2  Docs implemented;
     Changelog started;

0.3a Fix log filename;
     Remove configs;
     Fix config var;
     Implement pathlib;

0.3b Add ClamAV permissions denied handler;
     Fix bad API key, received from settings.json (bad stored key);
     Minor fixes;

0.3c Fix non-detected output (like permissions errors);
     Exclude List (sqlite3 db, 'exclude.db')

0.3d Exclude List fix;
     Apply basic SQL security setting;

Future plans:
    Web GUI (django?)
    New API migration https://onlinehelp.opswat.com/mdcloud/6.1_V4_Migration_guide.html
    


envy_sec.py
import argparse
import ipaddress
import logging
import os
import sys

try:
    from modules import clamav
    from modules import metadefender
    from modules import envy_settings
except (ModuleNotFoundError, ImportError):
    print('Failed to start secEnvyronment.')
    raise


class ConsoleInterface():
    """ envySec command line interface class.
    Used to perform scans and update from command-line.
    
    Receive Metadefender API key, ClamAV path to config.
    Required packages (dependencies): 
        inherit dependencies from ClamAV class and Metadefender Class
        and: ipaddress
    """

    def __init__(self, apikey = None, logging_level = 40):
        """ apikey -- Metadefender API key, see metadefender.py for more;
        logging_level -- verbosity of logging (0 - debug, 30 - warnings. See logging docs);
        """

        logging.basicConfig(level = logging_level,
                            filemode = 'a',
                            format='%(asctime)s >> %(name)s - %(levelname)s: %(message)s',
                            datefmt='%d.%m.%Y %H:%M:%S')
        
        self.envyCLI_Log = logging.getLogger('EnvySec CLI')
        self.envyCLI_Log.debug('__init__: Initializing class...')

        self.envy_conf = envy_settings.Envyronment_Settings(logging_level = logging_level)
        self.clam = clamav.ClamAV(self.envy_conf.clam_config, logging_level = logging_level)
        self.metadef = metadefender.Metadefender(self.envy_conf.settings["MetadefenderAPI"], logging_level = logging_level)

        self.envyCLI_Log.debug('__init__: Class initialized.')


    def ip_scanner(self, targets: list, geo = False) -> bool:
        """ Scan IP adress using Metadefender API.
        Backend defined at metadefender.py.
        
        Return dict with scan and geo data.
        Dict structure:
            scan_dump = {
                "1.2.3.4": {
                    "ScanData": [...],
                    "GeoData": [...]
                },
                "4.5.6.7": {
                    "ScanData": [...],
                    "GeoData": [...]
                },
                ...
        }

        For more details see official Metadefender documentations.
        """

        self.envyCLI_Log.debug('ip_scanner: Starting IP scan.')
        self.envyCLI_Log.debug('ip_scanner: received targets: {}'.format(targets))

        if type(targets) == str:
            targets = [targets]
        elif type(targets) == list and type(targets[0]) == list:
            targets = targets[0]

        self.envyCLI_Log.debug('ip_scanner: starting ip_scanner...')
        for target in targets:
            try:
                ip_addr = ipaddress.ip_interface(target)
            except ipaddress.AddressValueError as wrong_ip:
                print('envy_sec: Invalid IP address!')
                self.envyCLI_Log.error('ip_scanner: Invalid IP address: {}!'.format(targets))
                self.envyCLI_Log.debug('ip_scanner: ipaddress.AddressValueError args: {}'.format(wrong_ip))
                raise
            except ipaddress.NetmaskValueError as wrong_mask:
                print('envy_sec: Invalid IP mask!')
                self.envyCLI_Log.error('Iip_scanner: invalid IP mask: {}!'.format(targets))
                self.envyCLI_Log.debug('ip_scanner: ipaddress.NetmaskValueError args: {}'.format(wrong_mask))
                raise
            
            scan_dump = dict()
            for ip in ip_addr.network:
                ip = str(ip)

                self.envyCLI_Log.debug('ip_scanner: Gathering information for {}...'.format(ip))
                if ipaddress.ip_address(ip).is_global is True:
                    scan_data, geo_data = self.metadef.scan_ip(ip)

                    scan_dump[ip] = dict()
                    scan_dump[ip]['ScanData'] = scan_data
                    scan_dump[ip]['GeoData'] = geo_data
                    self.envyCLI_Log.info('ip_scanner: Gathering info for {} successfully done.'.format(ip))
                else:
                    self.envyCLI_Log.warning('Invalid IP address: {}!'.format(ip))
                    print('{} is not global IP, so not scanned.'.format(ip))
            
            self.envyCLI_Log.debug('ip_scanner: Calling for __show_ip_scan_results...')
            if self.__show_ip_scan_results(scan_dump, geo = geo) is True:
                self.envyCLI_Log.info('ip_scanner: Scanning {} is done.'.format(ip))
        
        self.envyCLI_Log.info('ip_scanner: Scan complete.')
        return True

    def __show_ip_scan_results(self, scan_results: dict, geo = True) -> bool:
        """ Parse data and print it to std.out. """

        try:
            for ip in scan_results:
                print('{} scan results:'.format(ip))
                for source in scan_results[ip]['ScanData']:
                    print('\t' + str(source) + ': ' + str(scan_results[ip]['ScanData'][source]))

                if geo is True:
                    print("""{} is placed in {}, {}, {}. 
                            Coordinates: {}, {}.""".format(ip, 
                                                            scan_results[ip]['GeoData']["Country"]), 
                                                            scan_results[ip]['GeoData']["Region"], 
                                                            scan_results[ip]['GeoData']["City"],
                                                            scan_results[ip]['GeoData']["Coordinates"]["Latitude"],
                                                            scan_results[ip]['GeoData']["Coordinates"]["Longitude"])
        except IndexError as ierr:
            self.envyCLI_Log.critical('__show_ip_scan_results: Unexpected index error.')
            self.envyCLI_Log.debug('__show_ip_scan_results: IndexError args: ' + str(ierr))
            raise
        else:
            self.envyCLI_Log.debug('__show_ip_scan_results: Parsing done successfully.')
            return True


    def file_scanner(self, targets: list) -> bool:
        """ Scan file. 
        
        Return True if scan complete successfully.
        Return False if file does not exist or not found.

        For more details see official Metadefender documentations.
        """

        self.envyCLI_Log.debug('file_scanner: Starting file scan.')
        self.envyCLI_Log.debug('file_scanner: received targets: ' + str(targets))

        print('Scanning...') # This may take a while...

        self.envyCLI_Log.debug('file_scanner:  checking variable type.')
        if type(targets) == str:
            targets = [targets]
        elif type(targets) == list and type(targets[0]) == list:
            targets = targets[0]
        self.envyCLI_Log.debug('file_scanner:  variable type checked.')

        self.envyCLI_Log.debug('file_scanner: starting file_scanner...')
        for target in targets:
            self.envyCLI_Log.debug('file_scanner: Start {} scan.'.format(target))
            if os.path.exists(target) is False:
                self.envyCLI_Log.error('file_scanner: {} does not exist or might not be accessed.'.format(target))
                print('\a' + str(target) + ' does not exist.')
                return False

        self.envyCLI_Log.debug('file_scanner: {} exist, starting ClamAV scanning.'.format(target))
        for i in self.clam.scan(targets):
            if str(i).endswith('FOUND\r') is True:
                i = i.split(': ')[0]

                self.envyCLI_Log.info('file_scanner: {} considered suspicious, starting Metadefender scan.'.format(target))
                self.envyCLI_Log.debug('file_scanner: Starting scanning by hash...')
                meta_response = self.metadef.scan_hash(i)
                if meta_response is False:
                    self.envyCLI_Log.debug('file_scanner: Hash scan failed, sending file...')
                    scan_result, scan_details = self.metadef.scan_file(i)
                    self.envyCLI_Log.debug('file_scanner: Response received, parsing...')
                    self.__parse_metadefender_scan(i, scan_result, scan_details)
                    self.envyCLI_Log.debug('file_scanner: Parsing complete.')
                else:
                    self.envyCLI_Log.debug('file_scanner: Hash scan successed.')
                    self.envyCLI_Log.debug('file_scanner: Response received, parsing...')
                    self.__parse_metadefender_scan(i, meta_response[0], meta_response[1])
                    self.envyCLI_Log.debug('file_scanner: Parsing complete.')
            else:
                self.envyCLI_Log.info('file_scanner: Unexpected behaviour: {}'.format(i))

        self.envyCLI_Log.debug('file_scanner: Scan complete.')
        return True

    def __parse_metadefender_scan(self, target: str, scan_result: dict, scan_details: dict) -> bool:
        """ Parse data and print it to std.out. """

        self.envyCLI_Log.debug('__parse_metadefender_scan: Starting parsing Metadefender response.')
        self.envyCLI_Log.debug('__parse_metadefender_scan: parsing response for ' + target + '.')

        print('Results for {}:'.format(target))
        print('\tTotal detections: {}'.format(scan_details["TotalDetections"]))
        for av in scan_result:
            print('\t\t{}: {}'.format(av, scan_result[av]))
        
        self.envyCLI_Log.debug('__parse_metadefender_scan: Parsing complete.')
        return True


    def add_exlcude(self, target: str) -> bool:
        """
        """

        self.clam.add_exception(target)
    
    def remove_exlcude(self, target: str) -> bool:
        """
        """

        self.clam.remove_exception(target)

    def get_exlcude(self) -> bool:
        """
        """

        for exception in self.clam.get_exception():
            print(exception)


    def update(self, verbose = False) -> bool:
        """ Simpe update command.
        Backend defined at clamav.py.

        Return True if update complete successfully.

        For update configuration details see config.json and freshclam.conf;
        For more details see official ClamAV documentations.
        """

        self.envyCLI_Log.debug('Signatures update started.')
        print('\tUpdating ClamAV signatures...')
        
        for update_output in self.clam.update():
            if verbose is True:
                print(update_output)
            else:
                pass

        self.envyCLI_Log.debug('Signatures update done.')
        print('\tDone.')
        return True


if __name__ == '__main__':

    logging.basicConfig(level = 0,
                        filename = os.path.abspath(os.path.dirname(__file__) + 'secEnvyronment.log'),
                        filemode = 'a',
                        format='%(asctime)s >> %(name)s - %(levelname)s: %(message)s',
                        datefmt='%d.%m.%Y %H:%M:%S')
    
    envy_sec = logging.getLogger('secEnvyronment')
    envy_sec.debug('Initialize Application...')

    parser = argparse.ArgumentParser(prog='secEnvyronment', epilog="""
                                    (c) EnvySec Dev, 2019.
                                    """, description="""
                                    This software is used to simplify ClamAV usage and automatize detect verification using Metadefender service.
                                    """)
    parser.add_argument('-F', '--scan-file', type=str, action='append', nargs='+', metavar='<PATH>',
                        help="""
                        Default scanning options will show only infected files.
                        Default path scanning behavior will recursively scan path.
                        If infected file found, it will be scanned using OPSWAT Metadefender.
                        
                        Example: envy_sec.py --scan-file C:\\* 
                            or envy_sec.py -S D:\\SomeFolder\\SomeFile.exe
                        """)
    parser.add_argument('-I', '--scan-ip', type=str, nargs='+', action='append',
                        metavar='<IPs>', help="""
                        IP will be scanned using OPSWAT Metadefender.
                        
                        Example: envy_sec.py --scan-ip 8.8.8.8 
                            or envy_sec.py -I 8.8.8.0\\24
                        """)
    parser.add_argument('-U', '--update', action='store_true', help="""
                        Update mode will update ClamAV Database.
                        (!) To Upgrade ClamAV itself download and install it manually.
                        
                        Example: envy_sec.py --update
                            or envy_sec.py -U
                        """)
    parser.add_argument('-E', '--add-exception', type=str, nargs='+', action='append',
                        metavar='<PATH>', help="""
                        Add path to exclude list. Files in exclude list will not be scanned by ClamAV.
                        
                        Example: envy_sec.py --add-exception C:\\
                            or envy_sec.py -E D:\\SomeFolder\\SomeFile.exe
                        """)
    parser.add_argument('-R', '--remove-exception', type=str, nargs='+', action='append',
                        metavar='<PATH>', help="""
                        Remove exception from exclude list.
                        
                        Example: envy_sec.py --remove-exception C:\\
                            or envy_sec.py -R D:\\SomeFolder\\SomeFile.exe
                        """)
    parser.add_argument('-G', '--get-exceptions', action='store_true', help="""
                        List all secEnvyronment exceptions.
                        
                        Example: envy_sec.py --get-exceptions
                            or envy_sec.py -G
                        """)
    parser.add_argument('-W', '--web', action='store_true', help="""
                        Start Web-based GUI.
                        
                        Example: envy_sec.py --web
                            or envy_sec.py -W
                        """)

    envy_sec.info('Parsing arguments...')
    args = parser.parse_args()
    envy_sec.debug('...parsing succeed.')
    
    if args.web == True:
        pass
    else:
        envy_sec.info('Initialize Command Line Interface (CLI).')
        envy_cli = ConsoleInterface() ## CLI initialize Metadefender and ClamAV itself.

        args.get_exception = True

        envy_sec.info('Initialize work:')
        if args.update is True:
            envy_sec.info('Starting update.')
            envy_cli.update(verbose = True)
            envy_sec.info('Update complete.')

        if args.scan_ip != None:
            envy_sec.info('Starting IP scan.')
            envy_cli.ip_scanner(args.scan_ip, geo = True)
            envy_sec.info('IP scan complete.')

        if args.scan_file != None:
            envy_sec.info('Starting file scan.')
            envy_cli.file_scanner(args.scan_file)
            envy_sec.info('File scan complete.')

        if args.add_exception != None:
            envy_sec.info('Adding exception to exclude list.')
            envy_cli.add_exlcude(args.scan_file)
            envy_sec.info('Exception added.')

        if args.remove_exception != None:
            envy_sec.info('Removing exception from exclude list.')
            envy_cli.remove_exlcude(args.scan_file)
            envy_sec.info('Exception removed.')

        if args.get_exceptions is True:
            envy_sec.info('Getting exceptions list.')
            envy_cli.get_exlcude()
            envy_sec.info('Exceptions list received.')

        if args.web is True:
            pass

    envy_sec.debug('secEnvyronment work complete.')